Wrappers/Python/cil/optimisation/utilities/sampler.py

```diff
@@ -0,0 +1,629 @@
+    >>> 
+    >>> #calculate prob weights 
+    >>> temp_list = []
+    >>> for i in range(num_indices):
+    >>>     temp_list.append(custom_list.count(i))
+    >>> total = sum(temp_list)
+    >>> prob_weights = [x/total for x in temp_list]
+    >>> 
+    >>> sampler = Sampler.from_function(num_indices=num_indices, function=test_function, prob_weights=prob_weights)
+    >>> print(list(sampler.get_samples(25)))
+    [1, 1, 1, 0, 0, 11, 5, 9, 8, 3, 1, 1, 1, 0, 0, 11, 5, 9, 8, 3, 1, 1, 1, 0, 0]
+    >>> print(sampler)
+    Sampler that wraps a function that takes an iteration number and selects from a list of indices {0, 1, …, S-1}, where S is the number of indices. 
+    Type : from_function 
+    Current iteration number : 11 
+    number of indices : 13 
+    Probability weights : [0.2, 0.3, 0.0, 0.1, 0.0, 0.1, 0.0, 0.0, 0.1, 0.1, 0.0, 0.1, 0.0] 
+
+
+
+    Note
+    -----
+    The optimal choice of sampler depends on the data and the number of calls to the sampler. 
+
+    For random sampling with replacement, there is the possibility, with a small number of calls to the sampler that some indices will not have been selected. For the case of uniform probabilities, the default, the number of
+    iterations required such that the probability that all indices have been selected at least once is greater than :math:`p` grows as :math:`nlog(n/p)` where `n` is `num_indices`. 
+    For example, to be 99% certain that you have seen all indices, for `n=20` you should take at least 152 samples, `n=50` at least 426 samples. To be more likely than not, for `n=20` you should take 78 samples and `n=50` you should take 228 samples. 
+    In general, we note that for a large number of samples (e.g. `>20*num_indices`), the density of the outputted samples looks more and more uniform. For a small number of samples (e.g. `<5*num_indices`) the user may wish to consider
+    another sampling method e.g. random without replacement, which, when calling `num_indices` samples is guaranteed to draw each index exactly once.  
+        """
+
+    def __init__(self, num_indices, function,  sampling_type=None, prob_weights=None):
+
+        self._type = sampling_type
+        self._num_indices = num_indices
+        self.function = function
+
+        if prob_weights is None:
+            prob_weights = [1/num_indices]*num_indices
+        else:
+            if abs(sum(prob_weights)-1) > 1e-6:
+                raise ValueError('The provided prob_weights must sum to one')
+
+            if any(np.array(prob_weights) < 0):
+                raise ValueError(
+                    'The provided prob_weights must be greater than or equal to zero')
+
+        self._prob_weights = prob_weights
+        self._iteration_number = 0
+
+    @property
+    def prob_weights(self):
+        return self._prob_weights
+
+    @property
+    def num_indices(self):
+        return self._num_indices
+
+    @property
+    def current_iter_number(self):
```

Comment:
I think this would be very difficult for the random samplers - to set the iteration number to 4000, one would need to sample up to and that number. Perhaps we just suggest that users redefine the samplers if they wisht to reset them?

Reply:
Maybe it's something we can add if requested, less for the stochastic but for deterministic it could be useful.